Creating a custom renderer

<a href="https://github.com/kiwistrongis"><img src="https://avatars.githubusercontent.com/u/2838640?" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [kiwistrongis](https://github.com/kiwistrongis)**
_Thursday Mar 13, 2014 at 00:42 GMT_
_Originally opened as https://github.com/processing/processing-android-archive/issues/64_

---

Moved from processing/processing#2390
Hi there, just working on something that will resolve the problems SMT was having with the latest version of processing - a custom renderer. I have it working fully on the desktop on Windows and Linux ( can't test on mac ). I ran into a problem on Android however.

This works on the desktop:

``` java
void setup(){
    size( 800, 600, Renderer.class.getName());
}
void draw(){
    background( 30);
    fill( 200, 100, 100, 200);
    noStroke();
    rect( 10, 10, 100, 100);
}
/*
public int sketchWidth() {
    return 800;
}
public int sketchHeight() {
    return 600;
}
public String sketchRenderer() {
    return Renderer.class.getName();
}
*/
public static class Renderer extends PGraphics3D {
    public Renderer(){
        super();}
}
```

However this fails on android:

``` java
void setup(){
    //size( 800, 600, Renderer.class.getName());
}
void draw(){
    background( 30);
    fill( 200, 100, 100, 200);
    noStroke();
    rect( 10, 10, 100, 100);
}

public int sketchWidth() {
    return 800;
}
public int sketchHeight() {
    return 600;
}
public String sketchRenderer() {
    return Renderer.class.getName();
}

public static class Renderer extends PGraphics3D {
    public Renderer(){
        super();}
}
```

With the following logcat output:

```
D/AndroidRuntime(17635): 
D/AndroidRuntime(17635): >>>>>> AndroidRuntime START com.android.internal.os.RuntimeInit <<<<<<
D/AndroidRuntime(17635): CheckJNI is OFF
D/AndroidRuntime(17635): Calling main entry com.android.commands.am.Am
D/AndroidRuntime(17635): Shutting down VM
D/AndroidRuntime(17649): 
D/AndroidRuntime(17649): >>>>>> AndroidRuntime START com.android.internal.os.RuntimeInit <<<<<<
D/AndroidRuntime(17649): CheckJNI is OFF
D/AndroidRuntime(17649): Calling main entry com.android.commands.pm.Pm
D/AndroidRuntime(17649): Shutting down VM
I/AndroidRuntime(17649): NOTE: attach of thread 'Binder Thread #3' failed
D/AndroidRuntime(17713): 
D/AndroidRuntime(17713): >>>>>> AndroidRuntime START com.android.internal.os.RuntimeInit <<<<<<
D/AndroidRuntime(17713): CheckJNI is OFF
D/AndroidRuntime(17713): Calling main entry com.android.commands.am.Am
D/AndroidRuntime(17713): Shutting down VM
I/AndroidRuntime(17713): NOTE: attach of thread 'Binder Thread #3' failed
D/AndroidRuntime(17741): Shutting down VM
E/AndroidRuntime(17741): FATAL EXCEPTION: main
E/AndroidRuntime(17741): java.lang.RuntimeException: Unable to start activity ComponentInfo{processing.test.asdf/processing.test.asdf.asdf}: java.lang.NullPointerException
E/AndroidRuntime(17741):    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:1960)
E/AndroidRuntime(17741):    at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:1985)
E/AndroidRuntime(17741):    at android.app.ActivityThread.access$600(ActivityThread.java:127)
E/AndroidRuntime(17741):    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1151)
E/AndroidRuntime(17741):    at android.os.Handler.dispatchMessage(Handler.java:99)
E/AndroidRuntime(17741):    at android.os.Looper.loop(Looper.java:137)
E/AndroidRuntime(17741):    at android.app.ActivityThread.main(ActivityThread.java:4447)
E/AndroidRuntime(17741):    at java.lang.reflect.Method.invokeNative(Native Method)
E/AndroidRuntime(17741):    at java.lang.reflect.Method.invoke(Method.java:511)
E/AndroidRuntime(17741):    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:784)
E/AndroidRuntime(17741):    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:551)
E/AndroidRuntime(17741):    at dalvik.system.NativeStart.main(Native Method)
E/AndroidRuntime(17741): Caused by: java.lang.NullPointerException
E/AndroidRuntime(17741):    at android.view.ViewGroup.addViewInner(ViewGroup.java:3342)
E/AndroidRuntime(17741):    at android.view.ViewGroup.addView(ViewGroup.java:3214)
E/AndroidRuntime(17741):    at android.view.ViewGroup.addView(ViewGroup.java:3184)
E/AndroidRuntime(17741):    at processing.core.PApplet.onCreate(Unknown Source)
E/AndroidRuntime(17741):    at android.app.Activity.performCreate(Activity.java:4466)
E/AndroidRuntime(17741):    at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1092)
E/AndroidRuntime(17741):    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:1924)
E/AndroidRuntime(17741):    ... 11 more
```

Help of any kind would be nice. Advice as to where to look and the like would be great.

--- a/core/src/processing/core/PApplet.java
+++ b/core/src/processing/core/PApplet.java
@@ -470,27 +470,30 @@ public class PApplet extends Activity implements PConstants, Runnable {
     //get renderer name anc class
     String renderer_name = sketchRenderer();
     Class<?> renderer_class = null;
-    try{
-      renderer_class = Class.forName( renderer_name);}
-    catch( ClassNotFoundException exception){
+    try {
+      renderer_class = Class.forName(renderer_name);
+    } catch (ClassNotFoundException exception) {
       //what to throw here?
-      throw new RuntimeException(
-        String.format(
-          "Error: Could not resolve renderer class name: %s", renderer_name),
-        exception);}
+      String message = String.format(
+        "Error: Could not resolve renderer class name: %s", renderer_name);
+      throw new RuntimeException(message, exception);
+    }
 
     //java2d renderer
-    if( renderer_name.equals( JAVA2D))
-      surfaceView = new SketchSurfaceView( this, sw, sh);
+    if (renderer_name.equals(JAVA2D)) {
+      surfaceView = new SketchSurfaceView(this, sw, sh);
+    }
     //P2D, P3D, and any other PGraphicsOpenGL-based renderer
-    else if( PGraphicsOpenGL.class.isAssignableFrom( renderer_class))
-      surfaceView = new SketchSurfaceViewGL( this, sw, sh,
+    else if (PGraphicsOpenGL.class.isAssignableFrom(renderer_class)){
+      surfaceView = new SketchSurfaceViewGL(this, sw, sh,
         (Class<? extends PGraphicsOpenGL>) renderer_class);
+    }
     //anything else
-    else throw new RuntimeException(
-      String.format(
-        "Error: Unsupported renderer class: %s",
-        renderer_name));
+    else {
+      String message = String.format(
+        "Error: Unsupported renderer class: %s", renderer_name)
+      throw new RuntimeException(message);
+    }
 
 //    g = ((SketchSurfaceView) surfaceView).getGraphics();
 
@@ -829,22 +832,26 @@ public class PApplet extends Activity implements PConstants, Runnable {
       // before surfaceChanged() is ever called.
 
       //P2D
-      if( renderer_class.equals( PGraphics2D.class))
+      if (renderer_class.equals( PGraphics2D.class)) {
         g3 = new PGraphics2D();
+      }
       //P3D
-      else if( renderer_class.equals( PGraphics3D.class))
+      else if (renderer_class.equals( PGraphics3D.class)) {
         g3 = new PGraphics3D();
-      //something that extends P2D or P3D
-      else
-        try{
+      }
+      //something that extends P2D, P3D, or PGraphicsOpenGL
+      else {
+        try {
           //dang java generics
           Constructor<? extends PGraphicsOpenGL> constructor =
             renderer_class.getConstructor();
-          g3 = constructor.newInstance();}
-        catch( Exception exception){
+          g3 = constructor.newInstance();
+        } catch (Exception exception) {
           throw new RuntimeException(
             "Error: Failed to initialize custom OpenGL renderer",
-            exception);}
+            exception);
+        }
+      }
 
 
       //set it up

